File: split-array-largest-sum.java
class Solution {
    public static int splitArray(int[] arr, int N, int K) {
        int low = Arrays.stream(arr).max().getAsInt(); // Minimum possible maximum subarray sum
        int high = Arrays.stream(arr).sum(); // Maximum possible maximum subarray sum

        while (low < high) {
            int mid = low + (high - low) / 2;

            if (isValid(arr, N, K, mid)) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }

        return low;
    }

    private static boolean isValid(int[] arr, int N, int K, int mid) {
        int count = 1;
        int sum = 0;

        for (int num : arr) {
            sum += num;

            if (sum > mid) {
                count++;
                sum = num;
            }
        }

        return count <= K;
    }
}

File: Difficulty: Hard/Split Array Largest Sum/split-array-largest-sum.cpp
class Solution {
  public:
    int splitArray(vector<int>& arr, int k) {
        int low = *max_element(arr.begin(), arr.end());
        int high = accumulate(arr.begin(), arr.end(), 0);
        int ans = high;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (isValid(arr, k, mid)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return ans;
    }

  private:
    bool isValid(vector<int>& arr, int k, int mid) {
        int count = 1, sum = 0;
        for (int num : arr) {
            if (sum + num > mid) {
                count++;
                sum = num;
                if (count > k) return false;
            } else {
                sum += num;
            }
        }
        return true;
    }
};
